core/src/com/unciv/logic/civilization/managers/UnitManager.kt

```diff
@@ -55,9 +55,11 @@ class UnitManager(val civInfo: Civilization) {
 
         val unit = civInfo.getEquivalentUnit(baseUnit)
         val citiesNotInResistance = civInfo.cities.filterNot { it.isInResistance() }
-        
+
+        // To allow cities on water tiles to be able to build water units.
+        val canSpwanUnitOnWater = city?.getCenterTile()?.isWater == true
         val cityToAddTo = when {
-            unit.isWaterUnit && (city == null || !city.isCoastal()) ->
+            unit.isWaterUnit && (city == null || !city.isCoastal() && !canSpwanUnitOnWater) ->
                 citiesNotInResistance.filter { it.isCoastal() }.randomOrNull() ?:
                 civInfo.cities.filter { it.isCoastal() }.randomOrNull()
             city != null -> city
```

Comment:
Wrong check - we need to add another branch here
- IF is water unit && current city can spawn - choose current city
- Else, if water unit, choose random water city (use isNaval not isCoastal)
- Else, if city, choose city
- Else, choose random city